### PR TITLE
Convert ContainsResultOperator to InExpression in relational

### PR DESCRIPTION
**Pull request by mikary**
Partial fix for #153
 - Resolves exception from invalid SQL and incorrect results for Contains subquery
 - Does not address issues with Contains subquery with nav props.
@anpete @maumar

### CODE REVIEW

File: src/EntityFramework.Relational/Query/Internal/RelationalResultOperatorHandler.cs [hunk 1]
```diff
@@ -213,7 +214,44 @@ private static Expression HandleCast(HandlerContext handlerContext)
...
         private static Expression HandleContains(HandlerContext handlerContext)
-            => handlerContext.EvalOnClient(requiresClientResultOperator: false);
+        {
+            var filteringVisitor
+                = handlerContext.SqlTranslatingExpressionVisitorFactory
+                    .Create(
+                        handlerContext.QueryModelVisitor,
+                        handlerContext.SelectExpression);
+
+            var item
```
Comment (maumar) [inline src/EntityFramework.Relational/Query/Internal/RelationalResultOperatorHandler.cs]: newItem

File: src/EntityFramework.Relational/Query/Internal/RelationalResultOperatorHandler.cs [hunk 2]
```diff
@@ -213,7 +214,44 @@ private static Expression HandleCast(HandlerContext handlerContext)
...
+                = filteringVisitor.Visit(
+                    ((ContainsResultOperator)handlerContext.ResultOperator).Item);
+
+            if (item != null)
+            {
+                var innerSelectExpression = handlerContext.SelectExpressionFactory.Create();
+
+                innerSelectExpression.Alias = "";
```
Comment (maumar) [inline src/EntityFramework.Relational/Query/Internal/RelationalResultOperatorHandler.cs]: add a comment why we need alias here, it's not obvious

File: test/EntityFramework.Core.FunctionalTests/QueryTestBase.cs [hunk 3]
```diff
@@ -1850,6 +1859,17 @@ where p.OrderDetails.Contains(od.FirstOrDefault(orderDetail => orderDetail.Disco
...
         }
 
+
+        [ConditionalFact]
+        public virtual void Where_subquery_on_collection()
+        {
+            AssertQuery<Product, OrderDetail>((pr, od) =>
+                pr.Where(
+                        p => od
+                            .Where(o => o.ProductID == p.ProductID)
```
Comment (maumar) [inline test/EntityFramework.Core.FunctionalTests/QueryTestBase.cs]: also add a test, that has a predicate on the OrderDetails, e.g. pr.Where( p => od.Where(o => o.Quantity > 7).Where(o => o.ProductID == p.ProductID).Select(o => o.Quantity).Contains<short>(5)))

File: src/EntityFramework.Relational/Query/Internal/RelationalResultOperatorHandler.cs [hunk 4]
```diff
@@ -213,7 +214,44 @@ private static Expression HandleCast(HandlerContext handlerContext)
...
+                        handlerContext.QueryModelVisitor,
+                        handlerContext.SelectExpression);
+
+            var item
+                = filteringVisitor.Visit(
+                    ((ContainsResultOperator)handlerContext.ResultOperator).Item);
+
+            if (item != null)
+            {
+                var innerSelectExpression = handlerContext.SelectExpressionFactory.Create();
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/Internal/RelationalResultOperatorHandler.cs]: Consider using SelectExpression.Clone

File: test/EntityFramework.Core.FunctionalTests/QueryTestBase.cs [hunk 5]
```diff
@@ -1841,8 +1841,17 @@ var efObjects
...
+        {
+            AssertQuery<Product, Product>((pr, pr2) =>
+                from p in pr
+                where pr2.Select(p2 => p2.ProductName).Contains("Chai")
+                select p);
+        }
+
         // TODO: [ConditionalFact] See #153
-        public virtual void Where_subquery_on_collection()
+        public virtual void Where_subquery_on_navigation()
```
Comment (anpete) [inline test/EntityFramework.Core.FunctionalTests/QueryTestBase.cs]: Move to nav query tests?

**Comment by anpete** [comment]
:shipit:

**Comment by maumar** [comment]
:shipit:

File: test/EntityFramework.Core.FunctionalTests/QueryTestBase.cs (hunk 3 above)
Comment (mikary) [inline test/EntityFramework.Core.FunctionalTests/QueryTestBase.cs]: Adding the predicate results in a query like
```sql
SELECT [p].[ProductID], [p].[Discontinued], [p].[ProductName], [p].[UnitsInStock]
FROM [Products] AS [p]
WHERE (
    SELECT CASE
        WHEN 5 IN (
            SELECT [o].[Quantity]
            FROM [Order Details] AS [o]
            WHERE ([o].[Quantity] > 7) AND ([o].[ProductID] = [p].[ProductID])
        )
        THEN CAST(1 AS BIT) ELSE CAST(0 AS BIT)
    END
) = 1
```
It looks like the SQL is right, but the composition of the predicates isn't part of this code change.

**Comment by mikary** [comment]
Fixes #4302

File: test/EntityFramework.MicrosoftSqlServer.FunctionalTests/QuerySqlServerTest.cs [hunk 6]
```diff
@@ -4104,6 +4104,45 @@ public override void Select_correlated_subquery_ordered()
...
+FROM [Products] AS [p]
+WHERE (
+    SELECT CASE
+        WHEN 'Chai' IN (
+            SELECT [p2].[ProductName]
+            FROM [Products] AS [p2]
+        )
+        THEN CAST(1 AS BIT) ELSE CAST(0 AS BIT)
+    END
+) = 1",
```
Comment (divega) [inline test/EntityFramework.MicrosoftSqlServer.FunctionalTests/QuerySqlServerTest.cs]: Can we simplify this translation to just have 'Chai' IN with a subquery in the WHERE clause?

Comment (divega) [inline test/EntityFramework.MicrosoftSqlServer.FunctionalTests/QuerySqlServerTest.cs]: Ah, just saw you filed https://github.com/aspnet/EntityFramework/issues/4304.

**Pull request merged**